ipq40xx: add support for Aliyun AP4220

1980490718 opened the pull request:
Hardware specifications
-------------
- SoC       : Qualcomm IPQ4018
- RAM       : 256 MiB DDR3
- Flash     : 4 MiB SPI NOR (Winbond) + 128 MiB SPI NAND (Winbond)
- WLAN      : IPQ4018 built-in
  - 2.4 GHz : 2x2 MIMO WiFi4
  - 5 GHz   : 2x2 MIMO WiFi5
- Ethernet  : QCA8072 10/100/1000 Mbps 1x WAN (PoE); 1x LAN
- USB       : 1x 2.0
- UART      : 1x Rj45 port, 115200n8
- Buttons   : 1x Reset
- LEDs      : 1x Power (green)
              1x WiFi 2.4 GHz (green)
              1x WiFi 5 GHz (green)
- Power     : 802.3at PoE

Flash instructions
-------------
  1. Connect the router via serial port
  2. Keep pressing "@" until uboot is interrupted
  3. Download the initramfs image, rename it to initramfs.bin, host it with tftp server
  4. Run these commands: "tftpboot initramfs.bin && bootm"
  5. After openwrt boots up, use scp or luci to upload sysupgrade.bin to upgrade.

robimarko commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -170,6 +170,19 @@ define Device/alfa-network_ap120c-ac
+	SOC := qcom-ipq4018
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	DEVICE_PACKAGES += ipq-wifi-alibaba_ap4220 -ath10k-board-qca4019
```
Literaly no point in defining `-ath10k-board-qca4019`

1980490718 replied:
OK

1980490718 replied:
> Literaly no point in defining `-ath10k-board-qca4019`

Done!

robimarko commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -170,6 +170,19 @@ define Device/alfa-network_ap120c-ac
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	DEVICE_PACKAGES += ipq-wifi-alibaba_ap4220 -ath10k-board-qca4019
+	FILESYSTEMS := squashfs
```
Drop this

1980490718 replied:
OK

1980490718 replied:
> Drop this

Done!

1980490718 commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,398 @@
+	};
+
+	key_pins: key_pinmux {
+		gpio {
```
means to delete it?

robimarko replied:
I mean, get rid of the gpio subnode, just keep its content directly under the original node

1980490718 replied:
> No point in a subnode if you are just having one function, same goes for all

Mybe,OK.  :)

1980490718 commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -171,6 +172,11 @@ ipq40xx_setup_macs()
 		label_mac=$(mtd_get_mac_binary "ART" 0x1006)
 		;;
+	alibaba,ap4220)
+		wan_mac=$(mtd_get_mac_text product_info 0x40)
```
> Please use NVMEM for caldata and MAC setting

Here or dts?    :)

robimarko replied:
Is it ASCII?
If so, then look at:
https://github.com/openwrt/openwrt/commit/eacc4d8c9b31ac17df034d2140558bdaaa56c16b

1980490718 replied:
without =

1980490718 commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,398 @@
+		gpios = <1 GPIO_ACTIVE_HIGH>;
+		line-name = "enable USB2 power";
+		output-low;
+	};
```
> No reason for hog, just define it as a GPIO regulator and pass it to USB PHY

 It's fixed.

robimarko commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,391 @@
+			label = "reset";
+			gpios = <&tlmm 63 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+			debounce-interval = <60>;
```
Drop the debounce interval, its not needed since this is IRQ based

robimarko commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,391 @@
+	model = "Alibaba AP4220";
+
+	aliases {
+		ethernet0 = &gmac;
```
Considering you are setting the MAC anyway it literally makes no sense to have this alias nor set the label MAC to it

robimarko commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,391 @@
+		regulator-max-microvolt = <5000000>;
+		regulator-min-microvolt = <5000000>;
+		regulator-name = "reg_usb";
+		regulator-boot-on;
```
This is uneeded and completely misses the point of PHY driver enabling the regulator

robimarko commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,391 @@
+	pinctrl-0 = <&spi0_pins>;
+	pinctrl-names = "default";
+
+	cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>, <&tlmm 4 GPIO_ACTIVE_HIGH>;
```
You gotta set `num-cs = <2>`  to indicate that that there are 2 CS

robimarko commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,391 @@
+	keys {
+		compatible = "gpio-keys";
+
+		pinctrl-0 = <&key_pins>;
```
Has this been even compiled?

Since `key_pins` is not even defined

robimarko commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,391 @@
+	leds {
+		compatible = "gpio-leds";
+
+		pinctrl-0 = <&led_pins>;
```
Again, this is not even defined

robimarko commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,391 @@
+		enable-active-low;
+		gpios = <&tlmm 1 GPIO_ACTIVE_LOW>;
+
+		pinctrl-0 = <&reg_usb_pins>;
```
Again undefined

robimarko commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,391 @@
+	};
+
+	serial_pins: serial_pinmux {
+		mux_uart {
```
Seriously, again with subnode in a node for no reason, just remove the `mux_uart` subnode but leave its content under `serial_pins`

aiamadeus commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,369 @@
+
+	reg_usb: regulator-usb {
+		compatible = "regulator-fixed";
+		enable-active-low;
```
`enable-active-low` is deprecated.

1980490718 replied:
> `enable-active-low` is deprecated.

Are you suggesting that this child node should be removed to avoid duplication with the '`gpios = <&tlmm 1 GPIO_ACTIVE_LOW>'` below?

aiamadeus replied:
Yes, this property is useless.

aiamadeus commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -170,6 +170,18 @@ define Device/alfa-network_ap120c-ac
+define Device/alibaba_ap4220
+	$(call Device/FitImage)
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := Alibaba
```
This should be `Aliyun`

1980490718 replied:
Thank's for your suggestion.I will implement the changes while resolving the conflicts.

1980490718 commented:
Dear team members, I've implemented all requested revisions. 
Could you kindly review the updated version and confirm if any further adjustments are needed? 
My sincere apologies if this follow-up causes any inconvenience.
@hauke @robimarko @aiamadeus

aiamadeus commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -170,6 +170,18 @@ define Device/alfa-network_ap120c-ac
+	SOC := qcom-ipq4018
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	DEVICE_PACKAGES += ipq-wifi-aliyun_ap4220
```
`DEVICE_PACKAGES :=`

aiamadeus commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-ap4220.dts:
```diff
@@ -0,0 +1,369 @@
+	};
+
+	spi-nand@1 {
+		status = "okay";
```
It is not disabled and does not need to be "okay".

1980490718 commented:
Why did the robot add so many labels? 
Will these labels affect the merge?

1980490718 commented:
**I added the following code into dts.**
`...
			partition@180000 {
				label = "product_info";
				reg = <0x180000 0x10000>;
				read-only;

				nvmem-layout {
					compatible = "ascii-eq-delim-env";
					#address-cells = <1>;
					#size-cells = <1>;

					product_mac_addr: product_mac {
						compatible = "mac-base";
						#nvmem-cell-cells = <1>;
					};
				};
			};
...
&swport4 {
	status = "okay";

	label = "wan";
	phy-handle = <&ethphy3>;
	nvmem-cells = <&product_mac_addr 0>;
	nvmem-cell-names = "mac-address";
};

&swport5 {
	status = "okay";

	label = "lan";
	phy-handle = <&ethphy4>;
	nvmem-cells = <&product_mac_addr 1>;
	nvmem-cell-names = "mac-address";
};
...
&wifi0 {
	status = "okay";

	nvmem-cell-names = "mac-address", "pre-calibration";
	nvmem-cells = <&product_mac_addr 2>, <&precal_art_1000>;
	qcom,ath10k-calibration-variant = "Aliyun-AP4220";
};

&wifi1 {
	status = "okay";

	nvmem-cell-names = "mac-address", "pre-calibration";
	nvmem-cells = <&product_mac_addr 3>, <&precal_art_5000>;
	qcom,ath10k-calibration-variant = "Aliyun-AP4220";
};
...`
**But the following error occurred.**
[    1.100800] 0x000000170000-0x000000180000 : "ART"
[    1.105986] OF: Bad cell count for /soc/spi@78b5000/flash@0/partitions
[    1.110124] 0x000000180000-0x000000190000 : "product_info"
[    1.116690] OF: Bad cell count for /soc/spi@78b5000/flash@0/partitions
[    1.122489] 0x000000190000-0x0000001b0000 : "mtdoops"
...
[   19.467680] ath10k_ahb a800000.wifi: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-nvmem max-sta 32 raw 0 hwcrypto 1
[   19.630453] kmodloader: done loading kernel modules from /etc/modules.d/*
[   29.911699] platform c000000.switch: deferred probe pending
[   29.911755] platform c080000.ethernet: deferred probe pending

**However, the driver 'ascii-eq-delim-env' is specifically suitable for the “name=value“and “name2=value2” format, but my device's 'produc_info' partition contains no '=' characters.
As I explained previously...**
root@OpenWrt:/# hexdump -C dev/mtd8
00000000  72 6f 2e 79 75 6e 6f 73  2e 63 6c 6f 75 64 75 75  |ro.yunos.clouduu|
00000010  69 64 09 36 42 45 35 32  46 42 36 36 30 41 31 44  |id.6BE52FB660A1D|
00000020  45 31 34 39 39 46 35 31  34 35 46 36 41 35 39 44  |E1499F5145F6A59D|
00000030  41 34 36 0a 70 72 6f 64  75 63 74 2e 6d 61 63 09  |A46.product.mac.|
00000040  42 38 3a 36 41 3a 39 37  3a 41 36 3a 39 45 3a 35  |B8:6A:97:A6:9E:5|
00000050  30 0a 70 72 6f 64 75 63  74 2e 6e 61 6d 65 09 5a  |0.product.name.Z|
00000060  53 6c 6d 33 35 5a 74 6e  62 63 75 78 4a 30 78 6a  |Slm35ZtnbcuxJ0xj|
00000070  38 46 4a 0a 70 72 6f 64  75 63 74 2e 73 6e 09 41  |8FJ.product.sn.A|
00000080  31 31 38 34 38 30 30 30  35 33 38 0a 70 72 6f 64  |11848000538.prod|
00000090  75 63 74 2e 73 65 63 72  65 74 09 6a 38 32 6b 63  |uct.secret.j82kc|
000000a0  4a 54 6a 34 4c 70 4c 76  75 56 64 4a 58 71 71 37  |JTj4LpLvuVdJXqq7|
000000b0  6d 74 34 30 78 4d 30 33  4c 57 78 0a 70 72 6f 64  |mt40xM03LWx.prod|
000000c0  75 63 74 2e 6d 6f 64 65  6c 09 41 50 34 32 32 30  |uct.model.AP4220|
000000d0  0a 70 72 6f 64 75 63 74  2e 76 65 6e 64 6f 72 09  |.product.vendor.|
000000e0  41 6c 69 62 61 62 61 20  47 72 6f 75 70 20 48 6f  |Alibaba Group Ho|
000000f0  6c 64 69 6e 67 20 4c 69  6d 69 74 65 64 0a ff ff  |lding Limited...|
00000100  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|

@robimarko

1980490718 committed:
ipq40xx: add support for Aliyun AP4220

Hardware specifications
-------------
- SoC       : Qualcomm IPQ4018
- RAM       : 256 MiB DDR3
- Flash     : 4 MiB SPI NOR (Winbond) + 128 MiB SPI NAND (Winbond)
- WLAN      : IPQ4018 built-in
  - 2.4 GHz : 2x2 MIMO WiFi4
  - 5 GHz   : 2x2 MIMO WiFi5
- Ethernet  : QCA8072 10/100/1000 Mbps 1x WAN (PoE); 1x LAN
- USB       : 1x 2.0
- UART      : 1x Rj45 port, 115200n8
- Buttons   : 1x Reset
- LEDs      : 1x Power (green)
              1x WiFi 2.4 GHz (green)
              1x WiFi 5 GHz (green)
- Power     : 802.3at PoE

Flash instructions
-------------
  1. Connect the router via serial port
  2. Keep pressing "@" until uboot is interrupted
  3. Download the initramfs image, rename it to
     initramfs.bin, host it with tftp server
  4. Run these commands:
     "tftpboot initramfs.bin && bootm"
  5. After openwrt boots up, use scp or luci
     to upload sysupgrade.bin to upgrade.

Signed-off-by: Willem Lee <1980490718@qq.com>
